### PR TITLE
#1: Add initial implementation of slc ci project

### PR DESCRIPTION
**Pull request by tomuben**

### CODE REVIEW

File: scripts/build/setup_poetry_env.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -o errexit
+set -o nounset
+set -o pipefail
+
+PYTHON_VERSION=$1
+
+PYTHON_BIN=$(command -v "$PYTHON_VERSION")
+poetry env use "$PYTHON_BIN"
+poetry install
```
Comment (Nicoretti) [inline scripts/build/setup_poetry_env.sh:10]: missing newline

File: test/test_branches.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                           target_docker_tag_prefix=test_env.commit_sha,
+                           target_docker_username=test_env.docker_user, target_docker_password=test_env.docker_pwd)
+        push_call_2 = call(push, flavor_path=("flavors/TEST_FLAVOR",), push_all=True, force_push=True, workers=7,
+                           target_docker_repository_name=test_env.docker_build_repo,
+                           target_docker_tag_prefix="",
+                           target_docker_username=test_env.docker_user, target_docker_password=test_env.docker_pwd)
+        assert (click_stub.invoke.mock_calls == [build_call, build_test_container_call,
+                                                 run_db_test_call_1, run_db_test_call_2,
+                                                 security_scan_call,
+                                                 push_call_1, push_call_2])
```
Comment (tkilias) [inline test/test_branches.py:66]: I like this

File: test/test_tests.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from exasol_integration_test_docker_environment.cli.commands.build_test_container import build_test_container
+from exasol_script_languages_container_tool.cli.commands import build, run_db_test, push, security_scan
+
+from script_languages_container_ci.lib.ci import ci
+
+from test.fixtures import build_output_dir, click_stub, test_env
+
+
+class TestTests:
+    @patch('script_languages_container_ci.lib.get_last_commit_message', MagicMock(return_value="Please be so kind and skip tests!"))
```
Comment (tkilias) [inline test/test_tests.py:13]: fancy

File: script_languages_container_ci/lib/__init__.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+def get_last_commit_message():
+    """
+    Assumes that PWD belongs to a GIT repository. Get's the last commit message of this repo and returns it as string
+    :return: Last commit message of current working directory GIT repository.
+    """
+    repo = Repo()
+    return repo.head.commit.message
+
+
+def print_file(filename: Path):
```
[the remark this answers is not in the document]
Comment (tomuben) [reply script_languages_container_ci/lib/__init__.py:36]: Remember all this stuff runs on AWS. For convenience we print this file in the log to find possible issues faster.

**Review by tkilias — CHANGES_REQUESTED** [review]

File: script_languages_container_ci/lib/__init__.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    logging.info(
+            """
+==========================================================
+Printing docker images
+==========================================================
+"""
+    )
+    for exa_img in exa_images:
+        logging.info(exa_img)
+    docker_client.close()
```
[the remark this answers is not in the document]
Comment (tomuben) [reply script_languages_container_ci/lib/__init__.py:24]: Done

File: script_languages_container_ci/lib/ci_build.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    """
+
+    logging.info(f"Running command 'build' with parameters: {locals()}")
+    ctx.invoke(build, flavor_path=flavor_path, force_rebuild=rebuild,
+               source_docker_repository_name=build_docker_repository,
+               source_docker_username=docker_user, source_docker_tag_prefix=commit_sha,
+               source_docker_password=docker_password, shortcut_build=False, workers=7)
+    logging.info(f"Running command 'build_test_container' with parameters: {locals()}")
+    ctx.invoke(build_test_container, force_rebuild=rebuild, workers=7)
+    print_docker_images()
```
[the remark this answers is not in the document]
Comment (tomuben) [reply script_languages_container_ci/lib/ci_build.py:29]: I separated reading and printing of the docker images, but will keep the `print_docker_images()`. It's more convinient.

File: .github/workflows/release_droid_upload_github_release_assets.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      - name: Build Release
+        run: bash ./scripts/build/build_release.sh "python3.8"
+      - name: Generate sha512sum files
+        run: find ./dist -maxdepth 1 -type f -exec bash -c 'sha512sum {} > {}.sha512' \;
+        shell: bash
+      - name: Upload assets to the GitHub release draft
+        uses: shogo82148/actions-upload-release-asset@v1
+        with:
+          upload_url: ${{ github.event.inputs.upload_url }}
+          asset_path: dist/*
```
[the remark this answers is not in the document]
Comment (tomuben) [reply .github/workflows/release_droid_upload_github_release_assets.yml:52]: Done

File: pyproject.toml [hunk 8]
```diff
@@ -12,15 +12,22 @@ authors = [
...
+pytest = "^7.1.1"
+pytest-mock = "^3.7.0"
+
+
+[tool.pytest.ini_options]
+minversion = "6.0"
+
+testpaths = [
+    "test"
+]
```
[the remark this answers is not in the document]
Comment (tomuben) [reply pyproject.toml:33]: Done

File: script_languages_container_ci/lib/__init__.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import logging
+from pathlib import Path
+
+import docker
+from git import Repo
+
+
+def print_docker_images():
```
[the remark this answers is not in the document]
Comment (tomuben) [reply script_languages_container_ci/lib/__init__.py:8]: I pass now logging.info as parameter.

File: script_languages_container_ci/lib/ci.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    Run CI build:
+    1. Build image
+    2. Run db tests
+    3. Run security scan
+    4. Push to docker repositories
+    """
+    logging.info(f"Running CI build for parameters: {locals()}")
+
+    rebuild = False
```
[the remark this answers is not in the document]
Comment (tomuben) [reply script_languages_container_ci/lib/ci.py:29]: Much better, indeed!

File: scripts/build/check_release.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+from git import Repo
+import toml
+
+
+def get_git_version():
+    repo = Repo()
+    assert not repo.bare
+    tag_strings = [t.name for t in repo.tags]
+    tag_strings.sort(reverse=True)
```
[the remark this answers is not in the document]
Comment (tomuben) [reply scripts/build/check_release.py:12]: Very cool!

File: script_languages_container_ci/lib/ci_security_scan.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+def ci_security_scan(ctx: click.Context, flavor_path: Tuple[str, ...]):
+    """
+    Run security scan and print result
+    """
+
+    logging.info(f"Running command 'security_scan' with parameters {locals()}")
+    ctx.invoke(security_scan, flavor_path=flavor_path, workers=7)
+    print_docker_images()
+    logging.info("============= SECURITY REPORT ===========")
+    print_file(Path() / ".build_output" / "security_scan" / "security_report")
```
[the remark this answers is not in the document]
Comment (tomuben) [reply script_languages_container_ci/lib/ci_security_scan.py]: I think we should change this, would be better if `exaslct` itself prints the report also to the logger.

File: pyproject.toml [hunk 13]
```diff
@@ -1,5 +1,5 @@
 [tool.poetry]
-name = "slc-ci"
+name = "script-languages-container-ci"
```
[the remark this answers is not in the document]
Comment (tomuben) [reply pyproject.toml]: Done

File: scripts/build/setup_poetry_env.sh (hunk 1 above)
Comment (tomuben) [reply scripts/build/setup_poetry_env.sh]: Done

File: test/test_tests.py [hunk 14]
```diff
@@ -1,44 +1,44 @@
...
+import pytest
 
 from exasol_script_languages_container_ci.lib.ci import ci
+from test import exaslct_calls
 
-from test.fixtures import tmp_test_dir, click_stub, test_env, patch_printfile
+from test.fixtures import tmp_test_dir, click_stub, patch_printfile
+from test.test_env import test_env
 
+testdata = [
```
Comment (tkilias) [inline test/test_tests.py:11]: I think, we should explain how the test data is structured

File: test/test_branches.py [hunk 15]
```diff
@@ -18,132 +17,55 @@ def last_commit():
...
-        """
-        Test that on branch master, we force rebuild and push also to release docker repository
-         1. Build Image (with force_rebuild)
-         2. Run db tests
-         3. Security scan
-         4. Push to docker build repo (with and withou sha)
-         5. Push to docker release repo
-        """
-        TEST_BRANCH = "refs/heads/master"
+testdata = [
```
Comment (tkilias) [inline test/test_branches.py:25]: I think, we should explain how the test data is structured

**Review by tkilias — CHANGES_REQUESTED** [review]

File: test/test_branches.py (hunk 15 above)
Comment (tomuben) [reply test/test_branches.py:25]: Done

File: test/test_tests.py (hunk 14 above)
Comment (tomuben) [reply test/test_tests.py:14]: Done

**Review by tkilias — DISMISSED** [review]

File: exasol_script_languages_container_ci/lib/__init__.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    return exa_images
+
+
+def print_docker_images(writer: Callable[[str], None]):
+    """
+    Prints all docker images whith "exa" in it's name to stdout.
+    :return: None
+    """
+
+    from inspect import cleandoc
```
Comment (Nicoretti) [inline exasol_script_languages_container_ci/lib/__init__.py:24]: This import I may would pull up to the top of the module. I know I put it there when I did the suggestion xD sorry :D.

File: exasol_script_languages_container_ci/lib/__init__.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    ))
+
+
+def get_last_commit_message():
+    """
+    Assumes that PWD belongs to a GIT repository. Get's the last commit message of this repo and returns it as string
+    :return: Last commit message of current working directory GIT repository.
+    """
+    repo = Repo()
+    return repo.head.commit.message
```
Comment (Nicoretti) [inline exasol_script_languages_container_ci/lib/__init__.py:40]: Assignment is only used once
```suggestion
    return Repo().head.commit.message
```

File: exasol_script_languages_container_ci/lib/__init__.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def print_file(filename: Path, writer: Callable[[str], None]):
+    """
+    Opens file readonly, reads it content and prints to writer.
+    """
+    with open(filename, "r") as f:
+        content = f.read()
+        writer(content)
```
Comment (Nicoretti) [inline exasol_script_languages_container_ci/lib/__init__.py:49]:
```suggestion
        writer(f.read())
```

**Review by Nicoretti — APPROVED** [review]

**Pull request merged**